ledger/src/blockstore.rs

```diff
@@ -3730,15 +3734,28 @@ fn is_orphan(meta: &SlotMeta) -> bool {
     meta.parent_slot.is_none()
 }
 
-// 1) Chain current_slot to the previous slot defined by prev_slot_meta
+// 1) Chain current_slot to the parent slot defined by parent_slot_meta
 // 2) Determine whether to set the is_connected flag
-fn chain_new_slot_to_prev_slot(
-    prev_slot_meta: &mut SlotMeta,
+fn chain_new_slot_to_parent_slot(
+    parent_slot_meta: &mut SlotMeta,
     current_slot: Slot,
     current_slot_meta: &mut SlotMeta,
 ) {
-    prev_slot_meta.next_slots.push(current_slot);
-    current_slot_meta.is_connected = prev_slot_meta.is_connected && prev_slot_meta.is_full();
+    let already_chained = parent_slot_meta.next_slots.contains(&current_slot);
+
+    // Ensure that next_slots never has duplicated slots
+    if already_chained {
+        debug_assert!(!already_chained, "slot was already chained to its parent");
+        datapoint_warn!(
+            "chained_duplicate_slot",
+            ("slot", current_slot, i64),
+            ("parent_slot", parent_slot_meta.slot, i64),
+        );
+    } else {
+        parent_slot_meta.next_slots.push(current_slot);
+    }
```

Comment:
so my inclination was that not to add this change of behavior until we can do some debugging to find the root cause. but your call.

Reply:
My hope is that this change treats one of the symptoms which might be all we need to keep things humming along when booting up from a snapshot. In the meantime, @willhickey is setting up canaries that we can enable assertions on for weird edge cases like this. If the assertion is hit, we can debug further. But unfortunately right now we don't have a way to repro this issue so this is the best I can come up with.